include/pingcap/kv/LockResolver.h

```diff
@@ -0,0 +1,142 @@
+{
+    uint64_t ttl;
+    uint64_t commit_ts;
+    bool isCommited() { return ttl == 0 && commit_ts > 0; }
+};
+
+constexpr size_t resolvedCacheSize = 2048;
+constexpr int bigTxnThreshold = 16;
+
+const uint64_t defaultLockTTL = 3000;
+
+const uint64_t maxLockTTL = 120000;
+
+// Lock represents a lock from tikv server.
+struct Lock
+{
+    std::string key;
+    std::string primary;
+    uint64_t txn_id;
+    uint64_t ttl;
+    uint64_t txn_size;
+
+    Lock(const ::kvrpcpb::LockInfo & l)
+        : key(l.key()), primary(l.primary_lock()), txn_id(l.lock_version()), ttl(l.lock_ttl()), txn_size(l.txn_size())
+    {}
+};
+
+using LockPtr = std::shared_ptr<Lock>;
+
+inline LockPtr extractLockFromKeyErr(const ::kvrpcpb::KeyError & key_err)
+{
+    if (key_err.has_locked())
+    {
+        return std::make_shared<Lock>(key_err.locked());
+    }
+    throw Exception("unknown error : " + key_err.ShortDebugString());
+}
+
+struct TxnExpireTime
+{
+    bool initialized;
+    int64_t txn_expire;
+
+    void update(int64_t lock_expire)
+    {
+        if (lock_expire <= 0)
+            lock_expire = 0;
+        if (!initialized)
+        {
+            txn_expire = lock_expire;
+            initialized = true;
+        }
+        else if (lock_expire < txn_expire)
+            txn_expire = lock_expire;
+    }
+
+    int64_t value() { return initialized ? txn_expire : 0; }
+};
+
+// LockResolver resolves locks and also caches resolved txn status.
+class LockResolver
+{
+public:
+    LockResolver(Cluster * cluster_) : cluster(cluster_), log(&Logger::get("pingcap/resolve_lock")) {}
+
+    // ResolveLocks tries to resolve Locks. The resolving process is in 3 steps:
+    // 1) Use the `lockTTL` to pick up all expired locks. Only locks that are too
+    //    old are considered orphan locks and will be handled later. If all locks
+    //    are expired then all locks will be resolved so the returned `ok` will be
+    //    true, otherwise caller should sleep a while before retry.
+    // 2) For each lock, query the primary key to get txn(which left the lock)'s
+    //    commit status.
+    // 3) Send `ResolveLock` cmd to the lock's region to resolve all locks belong to
+    //    the same transaction.
+
+    int64_t ResolveLocks(Backoffer & bo, uint64_t caller_start_ts, std::vector<LockPtr> locks);
+
+private:
+    void saveResolved(uint64_t txn_id, const TxnStatus & status)
+    {
+        std::unique_lock<std::shared_mutex> lk(mu);
+
+        if (resolved.find(txn_id) != resolved.end())
+            return;
+        cached.push(txn_id);
+        resolved.emplace(txn_id, status);
+        if (cached.size() > resolvedCacheSize)
+        {
+            auto to_remove = cached.front();
+            cached.pop();
+            resolved.erase(to_remove);
+        }
+    }
+
+    TxnStatus * getResolved(uint64_t txn_id)
+    {
+        auto it = resolved.find(txn_id);
+        if (it == resolved.end())
+            return nullptr;
+        return &(it->second);
```

Comment:
Is it thread safe to return a reference?

Reply:
Yes, the pointer is maintained by a global map and will not release until the map is full.